src/gt4py/next/iterator/type_inference.py

```diff
@@ -380,20 +385,54 @@ def _infer_shift_location_types(shift_args, offset_provider, constraints):
     return current_loc_in, current_loc_out
 
 
-class _TypeInferrer(eve.NodeTranslator):
+@dataclasses.dataclass
+class _TypeInferrer(eve.traits.VisitorWithSymbolTableTrait, eve.NodeTranslator):
     """Visit the full iterator IR tree, convert nodes to respective types and generate constraints."""
 
-    def __init__(self, offset_provider):
-        self.offset_provider = offset_provider
+    offset_provider: Optional[dict[str, Connectivity | Dimension]]
+    collected_types: dict[int, Type] = dataclasses.field(default_factory=dict)
+    constraints: set[tuple[Type, Type]] = dataclasses.field(default_factory=set)
 
-    def visit_SymRef(
-        self, node: ir.SymRef, constraints: set[tuple[Type, Type]], symtypes: dict[str, Type]
-    ) -> Type:
-        if node.id in symtypes:
-            res = symtypes[node.id]
-            if isinstance(res, LetPolymorphic):
-                return freshen(res.dtype)
-            return res
+    @classmethod
+    def apply(
+        cls,
+        node: ir.Node,
+        offset_provider: Optional[dict[str, Connectivity | Dimension]] = None,
+        reindex: bool = True,
+    ) -> dict[int, Type]:
+        """
+        Infer the types of the child expressions of a given iterator IR expression.
+
+        The result is a dictionary mapping the id of child nodes to their type.
+        """
+        # Collect preliminary types of all nodes and constraints on them
+        inferrer = cls(offset_provider=offset_provider)
+        inferrer.visit(node)
+
+        # Ensure dict order is pre-order of the tree
+        collected_types = dict(reversed(inferrer.collected_types.items()))
+
+        # Compute the most general type that satisfies all constraints
+        unified_types = unify(list(collected_types.values()), inferrer.constraints)
+
+        if reindex:
+            unified_types = reindex_vars(list(unified_types))
+
+        return {
+            id_: unified_type for id_, unified_type in zip(collected_types.keys(), unified_types)
+        }
+
+    def visit(self, node, **kwargs):
+        if isinstance(node, TYPED_IR_NODES):
+            self.collected_types[id(node)] = super().visit(node, **kwargs)
+            return self.collected_types[id(node)]
+
+        return super().visit(node, **kwargs)
```

Comment:
Am I missing something or this could be simplified to:
```suggestion
        result = super().visit(node, **kwargs)
        if isinstance(node, TYPED_IR_NODES):
            self.collected_types[id(node)] = result

        return result
```
?